livebook_examples/soundwave/soundwave.livemd

```diff
@@ -0,0 +1,244 @@
+>   An exemplary sample format might be `s16le` which stand for a signed integer written on 16 bits, with low endian order of bytes.
+> 
+> For some intuition on the formats you can take a look at a [`Membrane.RawAudio.SampleFormat` module](https://github.com/membraneframework/membrane_raw_audio_format/blob/master/lib/membrane_raw_audio/sample_format.ex)
+
+### Stream format handling
+
+Once the `stream_format` is received on the `:input` pad, some relevant information, i.e. the number of channels or the sampling rate, are fetched out of `Membrane.RawAudio` stream format structure. Based on that information a `VegaLite` chart is prepared.
+
+### Buffers handling
+
+Once a buffer is received, its payload is split into samples, based on `sample_format` of the `Membrane.RawAudio`. The amplitude of sound from different channels measured at the same time is average. As a result, a list of samples with each sample being an amplitude of sound at given time is produced.
+
+That list of samples is appended to the list of unprocessed samples stored in the element's state. Right after that `maybe_plot` function is invoked - and if there are enough samples, the samples are used to produce some points that are put on the plot.
+
+### Plotting of the soundwave
+
+Plotting all the audio samples with the typically used frequency (e.g. `44100 Hz`) is impossible due to limitations of the plot displaying system. That is why the list of samples is split into a number of chunks, and for each of these chunks a sample with `maximal` and `minimal` amplitude is found. For each chunk, only these two samples representing a given chunk are later put on the plot, with `x` value being a given sample timestamp, and `y` value being an measured amplitude of audio.
+tributes are used to drive the process of plotting:
+
+* `@windows_size` - describes the maximum number of points that are visible together on a plot,
+* `@window_duration` - describes the time range (in seconds) of points visible on the plot,
+* `@plot_updating_frequency` - describes how many times per second a plot should be updated with new points.
+  We encourage you to play with these attributes and adjust them to your needs. Please be aware, that setting to high `@windows_size` or `@plot_updating_frequency` might cause the plot to not be generated in realtime. At the same time, setting too low values of these parameters might result in loss of plot's accuracy (for instance making it insensitive to high frequency sounds).
+
+For more implementation details take a look at the code and the comments that describes parts, that might appear unobvious.
+
+```elixir
+defmodule Visualizer do
+  use Membrane.Sink
+
+  alias Membrane.RawAudio
+  alias VegaLite, as: Vl
+
+  @window_size 1000
+
+  # seconds
+  @window_duration 3
+
+  # Hz
+  @plot_update_frequency 50
+
+  def_input_pad(:input,
+    accepted_format: %RawAudio{},
+    demand_mode: :auto
+  )
```

Comment:
I believe the Livebook's formatter automatically adds these brackets ;/

Reply:
Hm, this is a `formatter.exs` from `membrane_core`
```elixir
locals_without_parens =
  [
    def_output_pad: 2,
    def_input_pad: 2,
    def_options: 1,
    def_clock: 1,
    def_type_from_list: 1,
    assert_receive_message: 3
  ]

[
  inputs: [
    "{lib,test,config,benchmark}/**/*.{ex,exs}",
    "mix.exs"
  ],
  locals_without_parens: locals_without_parens,
  export: [
    locals_without_parens: locals_without_parens
  ]
]
```
It might help you somehow resolve this problem

Reply:
If you cannot solve it, let it be as it is

Reply:
I think it's not possible to configure the livebook's formatter:
 https://elixirforum.com/t/how-to-configure-the-livebook-code-formatter/51323/4